### PR TITLE
Round of fixes for functional collectives

### PR DESCRIPTION
**Pull request by kumpera**
Move collective registration to torch.__init__ to handle multipy warmup. 
Fix all_reduce with non-contiguous tensors.

### CODE REVIEW

File: torch/__init__.py [hunk 1]
```diff
@@ -1489,3 +1489,10 @@ def _sparse_coo_tensor_unsafe(*args, **kwargs):
                   'use torch.sparse_coo_tensor(..., check_invariants=False) instead.')
     kwargs['check_invariants'] = False
     return torch.sparse_coo_tensor(*args, **kwargs)
+
+# Trigger registration of functional distributed collectives
+# This is needed by torch_deploy
+try:
+    import torch.distributed._functional_ops as funops
```
Comment (wconstab) [inline torch/__init__.py:1496]: do you want to bind these to a name (funops seems a bit playful) or do you just want to `import torch.distributed._functional_ops` here and not expose them under a short name

File: torch/__init__.py [hunk 2]
```diff
@@ -1489,3 +1489,10 @@ def _sparse_coo_tensor_unsafe(*args, **kwargs):
...
+except BaseException as _:
```
Comment (wconstab) [inline torch/__init__.py:1497]: So this isn't gonna be an ImportError? why BaseException

File: torch/distributed/_functional_collectives.py [hunk 3]
```diff
@@ -145,36 +147,25 @@ def _all_reduce(self, reduceOp, tag, ranks, group_size):
     group = c10d._find_or_create_pg_by_ranks_and_tag(tag, ranks, group_size)
     assert group is not None
 
-    inplace_tensor = self.clone()
+    inplace_tensor = self.clone(memory_format=torch.contiguous_format)
```
Comment (wconstab) [inline torch/distributed/_functional_collectives.py:150]: i _think_ the inductor version is already ok,  since it just creates a new 'FlexibleLayout' for AllReduce and doesn't do anything morally equivalent of '.clone'

**Review by wconstab — APPROVED** [review]
seems ok to me, tagging @ezyang to take a look too

File: torch/__init__.py (hunk 2 above)
Comment (ezyang) [reply torch/__init__.py:1497]: Don't suppress like this. Figure out the condition in which this module can be imported and do it that way

**Review by ezyang — DISMISSED** [review]
try except BaseException suppress is a big no no

File: torch/__init__.py (hunk 1 above)
Comment (kumpera) [reply torch/__init__.py:1496]: I'll `del funops` to ensure nothing gets imported into the `torch` module..

File: torch/__init__.py (hunk 2 above)
Comment (kumpera) [reply torch/__init__.py:1497]: 👍

File: torch/distributed/_functional_collectives.py (hunk 3 above)
Comment (kumpera) [reply torch/distributed/_functional_collectives.py:150]: We ran into this issue when running our e2e model.

File: torch/__init__.py [hunk 4]
```diff
@@ -1489,3 +1489,11 @@ def _sparse_coo_tensor_unsafe(*args, **kwargs):
                   'use torch.sparse_coo_tensor(..., check_invariants=False) instead.')
     kwargs['check_invariants'] = False
     return torch.sparse_coo_tensor(*args, **kwargs)
+
+# Trigger registration of functional distributed collectives
+# This is needed by torch_deploy
+try:
+    import torch.distributed._functional_ops as funops
+    del funops
+except ImportError as _:
```
Comment (Skylion007) [inline torch/__init__.py:1498]:
```suggestion
except ImportError:
```

Comment (kumpera) [reply torch/__init__.py]: I moved away from changing torch.__init__

File: torch/__init__.py (hunk 2 above)
Comment (kumpera) [reply torch/__init__.py]: Removed the import from `torch`

**Review by wconstab — APPROVED** [review]
ok looks good, we can unblock deploy later
